### PR TITLE
Fix navigate regions backwards for macOS Firefox and Safari.

### PR DESCRIPTION
**Pull request by afercia**
Noticed while testing https://github.com/WordPress/gutenberg/pull/44883

## What?

When pressing the keys Shift + Control + backtick to navigate backwards through the editor regions, the returned `event.key` varies across macOS browsers:
- Chrome: event.key is the backtick 
- Firefox: event.key is the tilde `~`
- Safari: event.key is the tilde `~`

My environment:
- macOS Monterey 12.6
- Firefox 105.0.3
- Safari Version 16.0
- Keyboard layout set to English

This makes region navigation fail on macOS Firefox and Safari, as the event.key is expected to match the backtick defined in the editor keyboard shortcuts.
 
## Why?

Navigation through the editor regions is broken and needs to be fixed.

## How?

Instead of trying to fix browser inconsistencies, seems to me the quickest way to fix this is by defining one more shortcut alias for the `Navigate to the previous part of the editor` command.

## Testing Instructions

**Test with Firefox or Safari on macOS. Make sure to test with an keyboard with an English layout or set your keyboard layout to English, as the backtick key may not be present on other keyboard layouts.**

Before:
- Go to the Post editor or to the Site editor.
- Click on any empty part of the interface to make sure the editor has focus.
- Press the key combination Shift + Control + backtick
- Observe that nothing happens.

After:
- Repeat the steps above.
- Observe the keyboard shortcut works as expected, moving focus to the previous editor region.
- Test also with Chrome, to make sure everything works as expected.

### CODE REVIEW

**Comment by ciampo** [comment]
Hey @afercia , could you also add an entry to the components package [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) for this fix? Thank you!

Also: is this MacOS specific behaviour? Do these changes cause any unintended behaviour in other OS's? Should we add unit tests for them?

**Comment by alexstine** [comment]
This has also been an issue for Windows. I just tested it in Windows with Firefox and Google Chrome, the issue is now resolved.

I would be in favor of adding an E2E test for this to catch this bug in the future. Besides the relevant test case and the changelog entry, this is ✅ .

Thanks.

**Comment by afercia** [comment]
Not an expert but I'm not sure an E2E test would help. The E2E tests run on Chrome, where the bug can't be reproduced because Chrome returns (incorrectly?) the backtick character while Firefox and Safari return the tilde.

**Comment by talldan** [comment]
https://github.com/WordPress/gutenberg/pull/44631 will introduce the option of running new playwright end to end tests in firefox and webkit (safari-ish).

**Comment by afercia** [comment]
Testing a bit more in depth: the Gutenberg shortcuts use [isKeyboardEvent](https://github.com/WordPress/gutenberg/blob/bf4eddb29f4c23ce841076b63035350866e81bfe/packages/keycodes/src/index.js#L347). It's based on `event.key` which turns to be unreliable across browsers.

macOS Chrome: Control + Shift + backtick returns the backtick character - works
Windows Chrome: Control + Shift + backtick returns the tilde character - fails

Firefox both platforms:  Control + Shift + backtick returns the tilde character - fails

Safari: Control + Shift + backtick returns the tilde character - fails

Opera both platforms:
Control + Shift + backtick returns the backtick character - works
Control + backtick returns only the Control key: the backtick key is ignored - fails

The new alias will only mitigate the problem. For the future, maybe some refactoring to move away from `event.key` should be taken into consideration.

**Comment by ciampo** [comment]
> Testing a bit more in depth: the Gutenberg shortcuts use [isKeyboardEvent](https://github.com/WordPress/gutenberg/blob/bf4eddb29f4c23ce841076b63035350866e81bfe/packages/keycodes/src/index.js#L347). It's based on `event.key` which turns to be unreliable across browsers.

I can confirm the discrepancy (via this [useful visualizer](https://www.toptal.com/developers/keycode)):

| Chrome MacOS | Firefox MacOS |
|---|---|
|  Screenshot 2022-10-18 at 11 47 57 | Screenshot 2022-10-18 at 11 48 06 |
| `event.key` is `~` | `event.key` is `` ` `` |

I guess the core of the issue is that the "backtick + shift" combination can produce two different values for `event.key` depending on the browser.

Just to understand, what is the expected key combination for navigating backwards ? Ctrl+tilde, or ctrl+shift+backtick?

> For the future, maybe some refactoring to move away from event.key should be taken into consideration.

As highlighted above, this may be solved by using `event.code` instead of `event.key`

**Comment by ciampo** [comment]
> #44631 will introduce the option of running new playwright end to end tests in firefox and webkit (safari-ish).

Testing for these inconsistencies sounds like a great usage of that new feature! We should rebase and add e2e tests as soon as #44631 is merged.

**Comment by alexstine** [comment]
@afercia If I test latest trunk on Windows Google Chrome (latest), this currently does not work. The keyboard shortcut `ctrl+` will move forward, but ctrl+shift+` does nothing.

Is there a difference in the way Mac handles this as well?

Thanks.

**Comment by alexstine** [comment]
Okay, the related PR was merged. @afercia Do you still have time to rebase and add E2E for this?

**Comment by afercia** [comment]
> If I test latest trunk on Windows Google Chrome (latest), this currently does not work.

@alexstine as mentioned earlier, Chrome appears to return a different `event.key` on macOS and Windows:

> macOS Chrome: Control + Shift + backtick returns the backtick character - works
> Windows Chrome: Control + Shift + backtick returns the tilde character - fails

Adding the tilde alias should fix it, as a temporary workaround.

Re: E2E tests, I'm afraid I won't be able to dedicate time to that.

**Comment by alexstine** [comment]
@ciampo Just added a Playwright E2E for this. I know it is not very detailed, but I think it will work for us for the moment until other tests can be migrated over. Thoughts?

This is also my first Playwright E2E creation, hopefully it looks okay.

Refreshed the branch, fixed conflicts, should be set on this one.

Thanks.

**Comment by ciampo** [comment]
Thank you for adding the e2e tests! To double-check, I ran them against `trunk` and they still seem to pass.

I had a quick look at [the `pressKeyWithModifier` function](https://github.com/WordPress/gutenberg/blob/8e8a8c5f237da0c74cea37a6073fb4d00f1a6a0e/packages/e2e-test-utils-playwright/src/page-utils/press-key-with-modifier.ts), which uses [the `keyboard.press()`](https://playwright.dev/docs/api/class-keyboard#keyboard-press) function from Playwright, which internally uses the `KeyboardEvent.key` property when creating the event. This likely means that [the bug that this PR is trying to solve](https://github.com/WordPress/gutenberg/pull/45019#issuecomment-1282129909) is simply not triggered by the `keyboard.press()` function.

To sum it up: these tests are useful for checking that backtick-related shortcuts allow a user to navigate regions, but they are not able to catch regressions caused by how different browsers interpret the combination of the `crtl+shift+backtick` keys and assign it to the `event.key` prop.

I'm not a playwright expert and I'm not if there's a different way to simulate a keypress (cc @kevin940726 @talldan in case they have any suggestion?)

**Comment by kevin940726** [comment]
Yeah, it seems like it's not possible in Playwright or Puppeteer. They both normalize the key codes. 😞  The best option I can think of is to manually test the inconsistency depending on the browser.

```js
if (testInfo.project.name === 'chromium') {
  await page.keyboard.press('Control+Shift+`');
} else {
  await page.keyboard.press('Control+Shift+~');
}
```

FWIW, any other key combinations with `Control+Shift` suffer from the same bug. For instance, `Control+Shift+1` yields `1` in chrome but yields `!` in safari. Maybe we should instead use `event.code` in the `keycodes` package to better fix this inconsistency in the future. This means breaking change though and should be done in another PR.

**Comment by afercia** [comment]
Thanks for all the testing and feedback,

> we should instead use event.code in the keycodes package to better fix this inconsistency in the future. This means breaking change though and should be done in another PR.

I'd agree. Can we merge this PR's temporary workaround for now?

**Comment by kevin940726** [comment]
> Can we merge this PR's temporary workaround for now?

Sure! Do you want to adjust the test as per my suggestion above, or do you think it's sub-optimal and we should revisit it later?

**Comment by afercia** [comment]
Will look into the test.

**Comment by afercia** [comment]
Adjusted the test as suggested. Can confirm that, when removing the tilde alias, the test passes with Chromium and fails with Webkit and Firefox, as expected. 

To run this single test in 'headed' mode:
`npm run test:e2e:playwright -- --headed editor/various/a11y-region-navigation.spec.js`

**Review by alexstine — APPROVED** [review]
LGTM.

File: test/e2e/specs/editor/various/a11y-region-navigation.spec.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+		// Navigate to next/second region and check that we made it.
+		await page.keyboard.press( 'Control+`' );
+		const editorContent = page.locator(
+			'role=region[name="Editor content"i]'
+		);
+		await expect( editorContent ).toBeFocused();
+
+		// Navigate to previous/first region and check that we made it.
+		// Make sure navigating backwards works also witht he tilde character,
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/a11y-region-navigation.spec.js]: Nit: typo

```suggestion
		// Make sure navigating backwards works also with the tilde character,
```

File: test/e2e/specs/editor/various/a11y-region-navigation.spec.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		// as browsers interpret the combination of the crtl+shift+backtick keys
+		// and assign it to event.key inconsistently.
```
Comment (kevin940726) [inline test/e2e/specs/editor/various/a11y-region-navigation.spec.js:42]: Nit: Might be worth adding a link to this PR in the comment here. It's optional though as git blame will do the same 😅 .

**Review by kevin940726 — APPROVED** [review]
LGTM 👍 . Just some small nitpicks.

**Comment by afercia** [comment]
Thank you @kevin940726 for taking the time to have a look and share feedback.
